specs/agents/tracing-instrumentation-aws-lambda.md

```diff
@@ -0,0 +1,158 @@
+`faas.trigger.region` | e.g. `us-east-1` |  Use `record.awsRegion` | Parse region segment (4th) from `record.sns.topicArn`
+`faas.execution` | e.g. `someMessageId` | Use `record.messageId` | Use `record.sns.messageId`
+`context.message.queue` | e.g. `arn:aws:sqs:us-east-2:123456789012:my-queue` | Use `record.eventSourceArn` | Use `record.sns.topicArn`
+`context.message.age` | e.g. `3298` | Age of the message in milliseconds. `current_time` - `SentTimestamp`, if SentTimestamp is available. </br></br> For SQS, the timestamp can be retrieved from message attributes with key `SentTimestamp`. | Age of the message in milliseconds. `current_time` - `SentTimestamp`, if SentTimestamp is available. </br></br> For SNS, the timestamp can be retrieved from `record.sns.timestamp`.
+`context.message.body` |  | The message body. Sould only be captured if body capturing is enabled in the configuration.</br></br> Use `record.body` | The message body. Sould only be captured if body capturing is enabled in the configuration.</br></br> Use `record.sns.message`
+`context.message.headers` |  | The message attributes. Should only be captured, if capturing headers is enabled in the configuration.</br></br> Use `record.messageAttributes`| The message attributes. Should only be captured, if capturing headers is enabled in the configuration.</br></br> Use `record.sns.messageAttributes`
+
+
+### S3
+Lambda functions that are triggered by S3 accept an `event` input that may contain one ore more `S3 event notification records` in the `event.records` array. We cannot (automatically) wrap the processing of the individual records that are sent as a batch of S3 event notification records with a single `event`. 
+
+Thus, in case that an S3 `event` contains **exactly one** `S3 event notification record`, the agents must apply the following, S3-specific retrieval of information. Otherwise, the agents should apply the [Generic Lambda Instrumentation](generic-lambda-instrumentation) as desribed above.
+
+In addition the following fields should be set for Lambda functions triggered by S3:
+Field | Value | Description 
+---   | ---   | ---
+`faas.trigger.type` | `datasource` | Constant value for message based triggers.
+`faas.trigger.name` | e.g. `mybucket` | Use `record.s3.bucket.name`
+`faas.trigger.id` | e.g. `arn:aws:s3:::mybucket/ObjectCreated:Put` | Use `record.s3.bucket.arn`
+`faas.trigger.region` | e.g. `us-east-1` | Use `record.awsRegion`
+`faas.trigger.version` | e.g. `2.1` | Use `record.eventVersion`
+`faas.execution` | e.g. `arn:aws:s3:::mybucket/ObjectCreated:Put` | Format: `${record.s3.bucket.arn}/${record.eventName}`
+
+
+## Init spans
+With both API Gateway versions the `event` objects (may) contain a timestamp of the original API Gateway request. The difference between the API Gateway timestamp and the actual start timestamp of the Lambda function is a good estimate for the initialization phase of the lambda function. 
```

Comment:
IIUC, queuing cannot happen, because in that case requests will be "throttled", which means they will fail with status code 429, as described here:
https://docs.aws.amazon.com/lambda/latest/dg/invocation-scaling.html

But I agree, that this is a sort of a workaround that is not precise. I'm also up for skipping the whole backdating part if it's misleading.
Thoughts on this?

Reply:
I would prefer to later improve our correlation with CloudWatch logs and metrics!
We can always go back to this and change it later too.

Reply:
Agreed. I've been struggling with whether backdating is valuable or not, as it skews all the metrics around the transaction lengths. And in the general case, a user is not doing things to affect the init time, but cares a lot about the time spent in the handler.

If we could pull init metrics out of cloudwatch, and limit the transaction to just the handler, I think that's a win-win.

Reply:
> they will directly fail with status code 429 if concurrency limit is reached

Ah, thanks.


I *think* that once we have a shared external extension that we can get a pretty good estimate of init phase time. Also if CloudWatch metrics provide these data, then we should look into slurping that data. (I see https://aws.amazon.com/kinesis/data-firehose/ mentioned for this kind of thing in some docs.)